library/src/main/java/co/fusionx/relay/util/Utils.java

```diff
@@ -10,55 +10,95 @@
 
 public class Utils {
 
+    static final char IRC_BOLD = '\u0002';
+    static final char IRC_COLOR = '\u0003';
+    static final char IRC_ITALIC = '\u001d';
+    static final char IRC_UNDERLINE = '\u001f';
+    static final char IRC_RESET_FORMATTING = '\u000f';
+
     public static boolean isNotEmpty(final CharSequence cs) {
         return !TextUtils.isEmpty(cs);
     }
 
     private static class FormatState {
+        private static final int NOT_OPEN = -1;
+
         private List<FormatSpanInfo> mFormats;
-        private int mBoldStart = -1, mColorStart = -1;
-        private int mItalicStart = -1, mUnderlineStart = -1;
+        private int mBoldStart = NOT_OPEN, mColorStart = NOT_OPEN;
+        private int mItalicStart = NOT_OPEN, mUnderlineStart = NOT_OPEN;
         FormatSpanInfo.Color mFgColor = null, mBgColor = null;
 
         void bold(int pos) {
-            if (mBoldStart < 0) mBoldStart = pos;
+            if (mBoldStart >= 0 && mBoldStart < pos) {
+                addBold(mBoldStart, pos);
+                mBoldStart = NOT_OPEN;
+            } else {
+                mBoldStart = pos;
+            }
         }
 
         void italic(int pos) {
-            if (mItalicStart < 0) mItalicStart = pos;
+            if (mItalicStart >= 0 && mItalicStart < pos) {
+                addItalic(mItalicStart, pos);
+                mItalicStart = NOT_OPEN;
+            } else {
+                mItalicStart = pos;
+            }
         }
 
         void underline(int pos) {
-            if (mUnderlineStart < 0) mUnderlineStart = pos;
+            if (mUnderlineStart >= 0 && mUnderlineStart < pos) {
+                addUnderline(mUnderlineStart, pos);
+                mUnderlineStart = NOT_OPEN;
+            } else {
+                mUnderlineStart = pos;
+            }
         }
 
         void color(int pos, FormatSpanInfo.Color fg, FormatSpanInfo.Color bg) {
             if (mColorStart >= 0 && pos > 0 && mFgColor != null) {
-                addFormat(new FormatSpanInfo(mColorStart, pos, mFgColor, mBgColor));
+               addColor(mColorStart, pos, mFgColor, mBgColor);
             }
             mColorStart = pos;
             mFgColor = fg;
             mBgColor = bg;
         }
 
-        void apply(int pos) {
+        void reset(int pos) {
```

Comment:
'reset' isn't really a good name for a method that essentially applies pending state, is it?